toolsrc/src/vcpkg/sourceparagraph.cpp

```diff
@@ -371,11 +372,14 @@ namespace vcpkg
             }
             else
             {
-                Debug::print("Failed to parse platform expression: ", opt.error(), "\n");
-                return nullopt;
+                r.add_generic_error(type_name(), opt.error());
+                return PlatformExpression::Expr::Empty();
```

Comment:
```suggestion
                return nullopt;
```

This is an error state, there's no reason to return a default constructed state.

Reply:
The reason to return a default constructed state is to indicate to the _parent_ code that the error has already been handled. Otherwise, you get two errors emitted for the same "problem": a specific error (emitted on the line above) and a generic error (emitted in the parent code).

With the previous JSON refactor, Deserializers should either:
1. Emit `nullopt` when they have no clue what's going on and don't have more specific error information to provide
2. Emit a more specific error with context information and return some default-constructed value.

Reply:
hmm, alright.

Reply:
That's an equivalent option, but I don't think it significantly improves the clarity of the code. It just pushes calling `r.add_error()` into the parent code, yet requires the formatting to be done in the child code. Since the context is omnipresent throughout the call stack, it's just as easy to simply call it directly and return a default value.

The further alternative would be to go all the way to some `ExpectedT<U, std::vector<string>>`, however this mandates a bunch of individual array allocations which makes me 😢.